docs/src/developer/typesystem.md

```diff
@@ -0,0 +1,217 @@
+of study and functions may be applied to maps to produce other maps.
+
+The simplest such function is composition. In a system where composition of
+maps always results in a map of the same type, no problem arises with the
+straightforward approach outlined above.
+
+However, for various reasons (including performance) it may not be desirable or
+even possible to construct a composition of two given maps using the same
+representation as the original maps. This means that the result of composing
+two maps of the same type may be a map of a different type, e.g. in the worst
+case a general composition type.
+
+This problem makes many homological and category theoretic operations on maps
+difficult or impossible to implement.
+
+Other operations which may be desirable to implement are caching of maps (e.g.
+where the map is extremely time consuming to compute, such as discrete
+logarithms) and attaching category theoretic information to maps. Such
+operations can be effected by encapsulating existing maps in objects containing
+the extra information, e.g. a cache or a category. However all the methods that
+applied to the original map objects now no longer apply to the encapsulated
+objects.
+
+To work around these limitations Nemo implements a four parameter `Map` type,
+`Map{D, C, T, U}`.
+
+The first two parameters are the domain and codomain types as discussed above.
+
+The parameter `T` is a "map class" which is itself an abstract type existing in
+a hierarchy of abstract types. This parameter is best thought of as a trait,
+independent of the hierarchy of abstract types belonging to `Map`, giving
+additional flexibility to the map types in the system.
+
+For example, `T` may be set to `LinearMap` or `FunctionalMap`. This may be
+useful if one wishes to distinguish maps in other ways, e.g. whether they are
+homomorphisms, isomorphisms, maps with section or retraction etc. As usual,
+offering traits partially gets around the single inheritance problem.
+
+The final parameter `U` is used to allow maps of a given type `U` to be
+composed and still result in a map of type `U`, even though the concrete type
+of the composition is different to that of the original maps. Methods can
+be written for all maps of type `U` by matching this parameter, rather than
+matching on the concrete type `U` of the original maps.
+
+For example, two maps with concrete type `MyRingHomomorphism` would belong to
+`Map{D, C, T, MyRingHomomorphism}` as would any composition of such maps, even
+if the concrete type of the composition was not a `MyRingHomomorphism`.
+
+Naturally four parameter types are rather unwieldy and so various helper
+functions are provided to compute four parameter map types. For example
+`Map(D, C)` (note the parentheses; it's a function which computes a type) will
+compute the union of all maps whose first two parameters are `D` and `C`, and
+where the remaining two parameters are arbitrary.
```

Comment:
What you describe seems to correspond precisely to `Map{D, C}` (with curly brackets); on the other hand, I don't find an actual `Map` method taking two arguments.

Reply:
Apparently you are right. Perhaps I never implemented that, or perhaps I removed it at some point. I've adjusted the text.